test/new-e2e/utils/e2e/e2e.go

```diff
@@ -0,0 +1,225 @@
+	credentialsManager := credentials.NewManager()
+	apiKey, err := credentialsManager.GetCredential(credentials.AWSSSMStore, "agent.ci.dev.apikey")
+	suite.require.NoError(err)
+
+	sshKey, err := credentialsManager.GetCredential(credentials.AWSSSMStore, "agent.ci.awssandbox.ssh")
+	suite.require.NoError(err)
+
+	suite.sshKey = sshKey
+	suite.add(config.DDAgentConfigNamespace, config.DDAgentAPIKeyParamName, apiKey)
+	suite.add(config.DDInfraConfigNamespace, aws.DDInfraDefaultKeyPairParamName, "agent-ci-sandbox")
+	suite.add(config.DDInfraConfigNamespace, aws.DDInfraDefaultInstanceTypeParamName, "t3.large")
+	suite.add(config.DDInfraConfigNamespace, aws.DDInfraDefaultARMInstanceTypeParamName, "m6g.medium")
+}
+
+func (c *Suite) add(namespace string, key string, value string) {
+	c.configMap[namespace+":"+key] = auto.ConfigValue{Value: value}
+}
+
+// TearDownSuite method is run after all the tests in the suite have been run.
+// This function is called by [testify Suite].
+//
+// [testify Suite]: https://pkg.go.dev/github.com/stretchr/testify/suite
+func (suite *Suite) TearDownSuite() {
+	if suite.fullStackName != "" {
+		var err error
+		if suite.destroyEnv {
+			err = infra.GetStackManager().DeleteStack(suite.ctx, suite.envName, suite.stackName)
+		}
+
+		if !suite.destroyEnv || err != nil {
+			stars := strings.Repeat("*", 50)
+			thisFolder := "A_FOLDER_CONTAINING_PULUMI.YAML"
+			if _, thisFile, _, ok := runtime.Caller(0); ok {
+				thisFolder = filepath.Dir(thisFile)
+			}
+
+			command := fmt.Sprintf("pulumi destroy -C %v --remove  -s %v", thisFolder, suite.fullStackName)
+			fmt.Fprintf(os.Stderr, "\n%v\nYour environment was not destroyed.\nTo destroy it, run `%v`.\n%v", stars, command, stars)
+		}
+		suite.require.NoError(err)
+	}
+}
+
+// Client provides a client to interact with the environment.
+type Client struct {
+	require   *require.Assertions
+	sshClient *ssh.Client
+}
+
+// Execute executes a command on the remote environment.
+func (c *Client) Execute(command string) string {
+	output, err := clients.ExecuteCommand(c.sshClient, command)
+	c.require.NoError(err, "Command `%v` failed with the output `%v`", command, output)
+	return output
+}
+
+// Run creates an environment and runs the E2E test.
+// The first argument is a function that creates the environment. It must return the type ([*vm.VM], error).
+// The second argument is a function that implements the E2E test logic.
+func (suite *Suite) Run(envFactory func(ctx *pulumi.Context) (*vm.VM, error), testFct func(*Client)) {
+	sshClient, err := suite.createEnv(envFactory)
+	suite.require.NoError(err)
+	defer sshClient.Close()
+	client := &Client{require: suite.require, sshClient: sshClient}
+	testFct(client)
+}
+
+func (suite *Suite) createEnv(fct func(ctx *pulumi.Context) (*vm.VM, error)) (*ssh.Client, error) {
+	stack, stackOutput, err := infra.GetStackManager().GetStack(suite.ctx, suite.envName, suite.stackName, suite.configMap, func(ctx *pulumi.Context) error {
```

Comment:
We seem to only return `stack` to get `stack.Name()`, isn't that equal to to `suite.stackName`?

Reply:
In the current implementation yes, but I don't want to assume that `stack.Name()` is equals to `suite.stackName` as this is an implementation detail of `GetStack`.